Hid dll

vbfox opened the pull request:
Add Hid.dll

Some notes :
- I added the new vaues to NTStatus before the name change should I rename them ?
- Same with Win32ErrorCode I parsed winerror and created the enum but converted the names to C#, should I rename them ?
- All Hid.Dll is documented as available in Universal, but I didn't test that
- No unit tests, pretty hard to test except the error conditions without forcing some Usb device to be present.

vbfox committed:
Add HIDP_ NTStatus members and al Win32 error codes

vbfox committed:
Add some functions of Hid.dll

AArnott commented on src/Windows.Core/NTStatusServerity.cs:
```diff
@@ -0,0 +1,16 @@
+    /// <summary>
+    /// Represent the 4 high bits of a <see cref="NTStatus"/> enumeration : Severity (2 bits), Customer code (1 bit) and Reserved (1 bit).
+    /// </summary>
+    internal static class NTStatusServerity
```
typo: Extra 'r' before the v.

vbfox replied:
Arf.. nice catch

AArnott commented on src/Windows.Core/NTStatusServerity.cs:
```diff
@@ -0,0 +1,16 @@
+    /// </summary>
+    internal static class NTStatusServerity
+    {
+        public const uint Success = (uint)0xC << 28;
```
Avoid using visibility modifiers that exceed the visibility of the containing type. Otherwise it looks like these contribute to the public API of the library when in fact they don't.

AArnott commented:
You'll likely have a merge conflict in NTStatus as I just changed it to add xml doc comments. Sorry about that.

AArnott commented:
> I added the new vaues to NTStatus before the name change should I rename them ?

Keep them exactly as they're defined in native code.

> Same with Win32ErrorCode I parsed winerror and created the enum but converted the names to C#, should I rename them ?

At this point, for consistency I think we should keep them as they are defined in native code. Otherwise NTStatus and Win32ErrorCode will be formatted differently and that will be weird. And we might discover we can't really maintain consistent renaming even within Win32ErrorCode for the same reason we can't for NTStatus.

> All Hid.Dll is documented as available in Universal, but I didn't test that

Assume the docs are correct, I guess, by making the methods available in the Shared project.

> No unit tests, pretty hard to test except the error conditions without forcing some Usb device to be present.

A unit test or two that tests a couple error conditions would be nice if it's not too hard. But it's not required.

AArnott commented:
Please make sure the build passes on AppVeyor and that you don't introduce new build warnings (I see a few StyleCop issues that should be producing warnings for you).

vbfox committed:
Fix the build

vbfox committed:
Fix StyleCop issues

vbfox commented:
> You'll likely have a merge conflict in NTStatus as I just changed it to add xml doc comments.

Anyway i'll need to change all the lines I added in there to keep the naming consistent

> At this point, for consistency I think we should keep them as they are defined in native code. Otherwise NTStatus and Win32ErrorCode will be formatted differently and that will be weird. And we might discover we can't really maintain consistent renaming even within Win32ErrorCode for the same reason we can't for NTStatus.

Ok, i'm still strongly in the camp of naming enum members as expected in any other C# lib. Even if it need work to solve conflict and need manual decisions instead of an automated process.
But you're the one maintaining the library on the long run, so i'll follow your conventions

> Please make sure the build passes on AppVeyor

I fucked up while trying to remember how nuget work (Using only paket theses days) and added a package that shouldn't be there, what I wonder is how it compiled on my work computer...

> and that you don't introduce new build warnings (I see a few StyleCop issues that should be producing warnings for you).

Yes it seem that the analyzer subsystem completely crashed on the current instance of VS on my work computer (No problem on my home one where I just pulled my branch), No warnings & no squiggles.

Only thing appearing is Resharper warning and as the solution doesn't have R# settings commited, they are partly false positives

vbfox committed:
Fix typo in NTStatusSeverity

vbfox committed:
Fix naming in Win32ErrorCode

vbfox committed:
Fix the build

vbfox committed:
Remove a wrapper that was checking exceptions

vbfox committed:
Use NTStatus.ThrowOnError() instead of manually throwing an exception in HidP_GetCaps

AArnott commented on src/Hid.Shared/Hid.cs:
```diff
@@ -0,0 +1,174 @@
+        /// </remarks>
+        [DllImport(nameof(Hid), CharSet = CharSet.Unicode, SetLastError = true)]
+        public static extern bool HidD_GetProductString(
+            SafeFileHandle hidDeviceObject,
```
Where is the SafeFileHandle type coming from?

AArnott replied:
When this type is no longer available (for portable), use the `SafeObjectHandle` type defined in Kernel32. You'll have to add a project reference to that from your desktop and portable projects, and add a project reference from you nuproj to the Kernel32 nuproj

AArnott commented on src/Hid.Shared/Hid+SafePreparsedDataHandle.cs:
```diff
@@ -0,0 +1,30 @@
+    /// </content>
+    public static partial class Hid
+    {
+        public class SafePreparsedDataHandle : SafeHandleMinusOneIsInvalid
```
Where is the `SafeHandleMinusOneIsInvalid` type coming from for the portable library build? I can't find it.

AArnott replied:
Ah, this too is because the .projitems import is missing from the portable library.

AArnott commented on src/Hid/Hid.csproj:
```diff
@@ -0,0 +1,41 @@
+      <Name>Windows.Core</Name>
+    </ProjectReference>
+  </ItemGroup>
+  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\Portable\$(TargetFrameworkVersion)\Microsoft.Portable.CSharp.targets" />
```
you seem to be missing an import:

```
<Import Project="..\Hid.Shared\Hid.Shared.projitems" Label="Shared" />
```

That would explain how you seem to be using types from Shared files that are not available in portable.

AArnott replied:
Ah, that's a bug in the template that was missing the tag. I've added it to the template. Please add it to this project file and fix up the build breaks that will probably follow.

AArnott commented on src/Hid.Shared/Hid.Helpers.cs:
```diff
@@ -0,0 +1,137 @@
+
+                default:
+                    result.ThrowOnError();
+                    return default(HidpCaps);
```
In the event that ThrowOnError doesn't throw (presumably because it was not in fact an error), do you really want to return an uninitialized struct, or would returning `hidCaps` be more appropriate?

vbfox replied:
Well if throw on error doesn't throw it mean that HidP_Caps returned a NtStatus of success, that it's not supposed to ever return (There is actually only 2 return value possible that are documented) so Ideally I should throw saying that nothing is as expected and that there might be some memory corruption for things to end there...

AArnott replied:
Sounds reasonable to do that then, rather than silently return an unexpected value.

AArnott commented on src/Hid.Shared/Hid.Helpers.cs:
```diff
@@ -0,0 +1,137 @@
+                    return hidCaps;
+
+                case NTStatus.HIDP_STATUS_INVALID_PREPARSED_DATA:
+                    throw new ArgumentException("The specified preparsed data is invalid.", nameof(preparsedData));
```
Is this case worth moving into the `ThrowOnError` method's switch statement so that this whole method can simply call `ThrowOnError` without a switch statement of its own?

vbfox committed:
Add some tests for Hid.dll

AArnott commented on src/Hid.Shared/Hid.Helpers.cs:
```diff
@@ -0,0 +1,137 @@
+
+        public static bool HidD_GetProductString(SafeFileHandle hidDeviceObject, out string result)
+        {
+            return GrowBuffer(sb => HidD_GetProductString(hidDeviceObject, sb, sb.Capacity), out result);
```
Do you want to make these helper methods return `string` and throw instead of return bool and use an out parameter?

vbfox replied:
I have the two versions as a version that doesn't throw but still handle the complexity of buffer growing could be useful (In my project I would replace the value with an empty string as they are only displayed to the user and some driver return errors as they don't support them)

AArnott commented on src/Hid.Shared/Hid.Helpers.cs:
```diff
@@ -0,0 +1,137 @@
+                    return false;
+                }
+
+                stringBuilder.Capacity = stringBuilder.Capacity * 2;
```
Should you cap this somehow so that you don't loop until you throw OutOfMemoryException? Is there some reasonable cut off beyond which you simply return the error code?

vbfox replied:
I could but it would be somehow arbitrary as anyway none of the HID protocols out there allow anything really big, what do you thing ? 1Mo ?

AArnott replied:
Sure.

AArnott commented on src/Hid.Shared/Hid+SafePreparsedDataHandle.cs:
```diff
@@ -0,0 +1,30 @@
+            public SafePreparsedDataHandle(IntPtr preexistingHandle, bool ownsHandle)
+                : base(ownsHandle)
+            {
+                this.handle = preexistingHandle;
```
tip: _maybe_ `SetHandle` is better to call here rather than setting the field directly. At the moment it appears they're equivalent, and maybe they'll always be. But since .NET itself sets it via the method, I suggest we do the same.

vbfox replied:
Yes I looked at the framework code for equivalent handles and will just do the same

AArnott commented:
BTW, I really like what you've done here. Your comments, and especially the trouble you took to use `<see/>` tags in your xml doc comments is first rate. :)

AArnott commented:
Oh, and feel free to add your name to the COPYRIGHT.md file if you'd like. This is a significant contribution. :)

vbfox committed:
Make the portable project build

vbfox committed:
Add missing nuproj reference to kernell32

vbfox committed:
Limit string grow to 1Mo in Hid helpers and Fix unit tests

vbfox committed:
Always throw in HidP_GetCaps if we don't have one of the two expected results

vbfox committed:
Add me (vbfox) to the COPYRIGHT.md

AArnott commented on src/Hid.Tests/Hid.cs:
```diff
@@ -0,0 +1,61 @@
+    [Fact]
+    public void HidD_GetAttributes_ThrowForInvalidHandle()
+    {
+        Assert.Throws<COMException>(() =>
```
Does throwing a COMException make sense? This isn't really COM we're dealing with here. But I'm not sure what better one to throw.

vbfox replied:
I don't like it either, really just did what ntstatus does. Win32Exception isn't in the portable profile so if we want to use it it'll need to be re-implemented somewhere.

AArnott replied:
Interesting. `Win32Exception`, with all its magic, is actually pretty simple. I suppose we could recreate it here (and probably should, given our scope). But in another PR.
